Add JSON streaming for RESTEasy Reactive Jsonb and Jackson

Sgitario opened the pull request:
We now support json streaming. Example:

```java
    @Path("stream-json/multi")
    @GET
    @Produces(SseMediaType.APPLICATION_STREAM_JSON)
    @RestSseElementType(MediaType.APPLICATION_JSON)
    public Multi<Message> multiStreamJson() {
        return Multi.createFrom().items(new Message("hello"), new Message("stef"));
    }
```

We support "application/x-ndjson" and "application/stream+json".

Moreover, I've added a new utility class called `SseMediaType` to be used by users. The implementation is similar to `MediaType`.

Questions:
- Added for both JSONB and Jackson extensions, is this correct?
- I've copied the existing tests we had for sse in JSONB to Jackson, is this correct?

Fix https://github.com/quarkusio/quarkus/issues/13663

Sgitario commented:
> PR needs formatting

It should be fixed now

Sgitario commented:
PR updated to resolve some new conflicts with main.

geoand commented:
This now needs a rebase

Sgitario commented:
> This now needs a rebase

Done

geoand commented:
@FroMage can you take a look at this one?

gsmet commented:
@geoand what should we do with this one?

geoand commented:
Ping @FroMage again 😊

FroMage commented on independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/util/SseMediaType.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Representation of the media types suitable for server streaming.
+ */
+public final class SseMediaType {
```
I disagree with that name, JSON streaming isn't SSE (Server-Sent Events) which is its own spec. We could however call it `StreamMediaType`.

Or better yet we make our own subtype of `MediaType` that is `RestMediaType` which has these additional types?

Sgitario replied:
Renamed to `StreamMediaType`. 
I didn't rename to `RestMediaType` because using `RestMediaType.APPLICATION_NDJSON` sounds weird to me. Maybe `ReactiveMediaType`?

FroMage replied:
And making it a subtype of `MediaType` allows us to just tell everyone to use `RestMediaType` for all their media types.

Sgitario replied:
Okis, will rename it to `RestMediaType`.

Sgitario replied:
Renamed to `RestMediaType`.

FroMage commented on extensions/resteasy-reactive/quarkus-resteasy-reactive-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/sse/SseResource.java:
```diff
@@ -0,0 +1,125 @@
+    @Path("ndjson/multi")
+    @GET
+    @Produces(SseMediaType.APPLICATION_NDJSON)
+    @RestSseElementType(MediaType.APPLICATION_JSON)
```
Now I wonder if we should not rename this `@RestStreamElementType`? Though to keep it compatible we'd keel `@RestSseElementType` and deprecate it in favour of the new one.

Sgitario replied:
I agree with you, however after taking a look to the required changes, there were many places to update within the resteasy reactive framework and I want to keep this PR the simpler the better. Wdyt?

FroMage replied:
Well, I'd rather group this from the start, if we can, since we're going to document this.

Sgitario replied:
Fine with me to do this change. fyi @geoand

Sgitario replied:
I've done this in a separated commit: https://github.com/quarkusio/quarkus/pull/20908/commits/0c102bf51abd4c70e663ff042a05522d2d05335f

FroMage reviewed (changes requested):
This looks good, except a few comments, and requires a mention in the documentation, linking to the stream specs.

FroMage commented:
Really sorry about the late review, this completely slipped past me.

Sgitario commented:
@FroMage PR updated with one of your suggestions

Sgitario committed:
Add JSON streaming for RESTEasy Reactive Jsonb and Jackson

We now support json streaming. Example:

```
@Path("stream-json/multi")
    @GET
    @Produces(SseMediaType.APPLICATION_STREAM_JSON)
    @RestSseElementType(MediaType.APPLICATION_JSON)
    public Multi<Message> multiStreamJson() {
        return Multi.createFrom().items(new Message("hello"), new Message("stef"));
    }
```

We support "application/x-ndjson" and "application/stream+json".

Moreover, I've added a new utility class called `SseMediaType` to be used by users. The implementation is similar to `MediaType`.

FroMage commented on independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/util/RestMediaType.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Extended media types in Resteasy Reactive.
+ */
+public final class RestMediaType {
```
Any reason why this is not a subtype of `MediaType`?

Sgitario replied:
Sorry, I've missed this change. PR updated!

FroMage reviewed (changes requested):
Looks great, just the question of the supertype of `RestMediaType` left.

Sgitario committed:
Deprecate @RestSseElementType in favour of @RestStreamElementType

FroMage reviewed (approved):
Looks great, thanks!

FroMage commented:
CI failure unrelated: merging
